### PR TITLE
feat: add postInitApplicationSQLRefs

### PR DESCRIPTION
**Pull request by Anhui-tqhuang**
Add postInitApplicationSQLRefs, as a result users could refer to post init application SQLs from 
files inside configmap or secrets.

Closes https://github.com/cloudnative-pg/cloudnative-pg/issues/250 

Signed-off-by: TQ [tianqiu.huang@enterprisedb.com](tianqiu.huang@enterprisedb.com)

### CODE REVIEW

File: internal/cmd/manager/instance/initdb/cmd.go
Comment (phisco) [inline internal/cmd/manager/instance/initdb/cmd.go:69]: Given that these secrets and configmaps are needed only in the init job we could think of mounting them directly in the pod, maybe in a single directory with an index (e.g. "00-....sql", 01-... etc) so that we can keep track of the ordering. This way we would avoid requiring permissions to access the configmaps and the secrets, a client and we could pass a flag specifying the folder only, instead of the whole list of secrets with keys. What do you think, @Anhui-tqhuang ?

Comment (Anhui-tqhuang) [reply internal/cmd/manager/instance/initdb/cmd.go:69]: yeah, that's a good point

File: pkg/specs/volumes.go
Comment (sxd) [inline pkg/specs/volumes.go:91]: magic numbers, the user needs to provide the volume for this, if the user wants and will know the size of the volume, other wise, we can be creating a volume that the user will know nothing about

File: pkg/specs/volumes.go
Comment (sxd) [inline pkg/specs/volumes.go:180]: again, the user doesn't know about this magic volume

File: api/v1/cluster_types.go
Comment (Anhui-tqhuang) [reply api/v1/cluster_types.go:763]: reference to secrets or configmaps

File: api/v1/cluster_webhook.go
Comment (Anhui-tqhuang) [reply api/v1/cluster_webhook.go:420]: if we don't have the secrets or configmaps ready, the init job will not be abled to be completed

File: api/v1/cluster_webhook.go
Comment (Anhui-tqhuang) [reply api/v1/cluster_webhook.go:430]: if we don't have the secrets or configmaps ready, the init job will not be abled to be completed

File: pkg/management/postgres/initdb.go
Comment (Anhui-tqhuang) [reply pkg/management/postgres/initdb.go]: gotcha, thanks!

File: pkg/specs/volumes.go
Comment (Anhui-tqhuang) [reply pkg/specs/volumes.go:91]: user doesn't provider volume, they provider refs to secrets or configmaps only, and we attach volumes to init job spec by the controller.

i assume user cannot change the spec of init job directly

File: pkg/specs/jobs.go
Comment (Anhui-tqhuang) [reply pkg/specs/jobs.go:73]: i follow the strategy used by other stuffs like `postInitApplicationSQL`

File: pkg/specs/volumes.go
Comment (Anhui-tqhuang) [reply pkg/specs/volumes.go:91]: also i want to keep the magic number here to make sure the order of implementation to sql files

**Comment by sxd** [comment]
@Anhui-tqhuang if this is a work in progress please set it as draft so we look after it's finished :)

**Comment by Anhui-tqhuang** [comment]
@sxd converted

**Comment by phisco** [comment]
🤩 awesome patch @Anhui-tqhuang! I will review and test it later more in detail, but at a quick glance it looks really well done! Thanks for your contribution!

**Comment by Anhui-tqhuang** [comment]
@sxd @phisco hello, the PR is ready, could i get a review from you plz

File: api/v1/cluster_webhook.go
Comment (sxd) [reply api/v1/cluster_webhook.go:420]: Then the init job should fail saying that the secret it's missing

File: pkg/specs/jobs.go
Comment (sxd) [reply pkg/specs/jobs.go:73]: that doesn't mean things can be improved...

File: pkg/specs/volumes.go
Comment (sxd) [reply pkg/specs/volumes.go:91]: nop, it's not acceptable to keep magic numbers, in that case a directory inside $PGDATA it's a better solution

File: api/v1/cluster_types.go
Comment (phisco) [inline api/v1/cluster_types.go:779]: please use our own SecretKeySelector and ConfigMapKeySelector, see https://github.com/Anhui-tqhuang/cloudnative-pg/blob/main/api/v1/base_types.go#L32.

File: pkg/specs/volumes.go
Comment (Anhui-tqhuang) [reply pkg/specs/volumes.go:101]: good point, will improve

File: pkg/specs/volumes.go
Comment (Anhui-tqhuang) [reply pkg/specs/volumes.go:210]: you are right

**Comment by Anhui-tqhuang** [comment]
@phisco hey i addressed all comments, and the PR is ready for another PASS

**Comment by phisco** [comment]
@Anhui-tqhuang can you run a `make checks` and push any changed file to solve https://github.com/cloudnative-pg/cloudnative-pg/runs/7017512429?check_suite_focus=true?

**Comment by phisco** [comment]
As I was saying the patch looks good, the only thing I wanted to improve was to find a way to let users know the initdb job is pending because of a missing configmap/secret we'd have to mount. I have a rough implementation at the moment, but i think we could address this more broadly, maybe we could discuss it with the rest of the team to get also their feedback and decide whether to move on this PR as is and address that later or use this pr for that too.

**Comment by gbartolini** [comment]
What's the status of this patch? I see it is not assigned to any milestone, but I see that 2 maintainers have already reviewed it. Why did not it land in 1.16?

**Comment by gbartolini** [comment]
I chatted with @phisco and he suggested we move this to 1.17 for now.

**Comment by mnencia** [comment]
This PR needs to be rebased on top of the offline database import patch, which also introduces the `PostImportApplicationSQL` field that is a good candidate for a `Refs` pair.

**Comment by Anhui-tqhuang** [comment]
> This PR needs to be rebased on top of the [offline data import patch](https://github.com/cloudnative-pg/cloudnative-pg/commit/1b63d6a3be2efe14e62d32ef7e38df268ef9fc41), which also introduces the PostImportApplicationSQL field that is a good candidate for a Refs pair.

@mnencia good point, i would like to add this in another PR to avoid making this PR so big

**Comment by Anhui-tqhuang** [comment]
@phisco hello, i rebased the branch and resolved the conflicts, could we move forward on this PR?

**Review by phisco — APPROVED** [review]

**Comment by phisco** [comment]
@Anhui-tqhuang :wave: we have discussed this patch in the latest community call, could you update the documentation around [here](https://github.com/cloudnative-pg/cloudnative-pg/blob/adf57a2340ec03315f35deba052eab4d89436cab/docs/src/bootstrap.md#bootstrap-an-empty-cluster-initdb) and maybe add an example in the examples folder similar to [this](https://github.com/cloudnative-pg/cloudnative-pg/blob/main/docs/src/samples/cluster-example-initdb.yaml). We agreed it's not t going to get into 1.16, but will be definitely merged for 1.17, and about what @mnencia was mentioning [here](https://github.com/cloudnative-pg/cloudnative-pg/pull/287#issuecomment-1173920449), we agreed it can be handled in a different patch, let us know if you want to address that too!

**Comment by Anhui-tqhuang** [comment]
@phisco hey, i added the doc and example in this [commit](https://github.com/cloudnative-pg/cloudnative-pg/pull/287/commits/8eafa229aff0002b27cdef146c5442f52d273fd3)

> and about what @mnencia was mentioning https://github.com/cloudnative-pg/cloudnative-pg/pull/287#issuecomment-1173920449, we agreed it can be handled in a different patch, let us know if you want to address that too!

yeah, i would like to address that too!

**Review by phisco — APPROVED** [review]

File: docs/src/bootstrap.md
Comment (sxd) [inline docs/src/bootstrap.md:278]: the proper words should be `Secrets` and  `ConfigMaps`

File: docs/src/bootstrap.md
Comment (sxd) [inline docs/src/bootstrap.md:307]: the proper words should be `Secrets` and `ConfigMaps`

**Comment by sxd** [comment]
@Anhui-tqhuang can you please add the sample file add into the samples.md file with the proper prerequisites too please

File: docs/src/bootstrap.md
Comment (Anhui-tqhuang) [reply docs/src/bootstrap.md:278]: done

File: docs/src/bootstrap.md
Comment (Anhui-tqhuang) [reply docs/src/bootstrap.md:307]: done

**Comment by Anhui-tqhuang** [comment]
@sxd thx for the review
> @Anhui-tqhuang can you please add the sample file add into the samples.md file with the proper prerequisites too please

i will add it to samples.md without the section `prerequisites` as my sample file already contains the definitions to Secrets and ConfigMaps

File: pkg/utils/digits.go
Comment (Anhui-tqhuang) [reply pkg/utils/digits.go:20]: > return len(fmt.Sprintf("%d", n))

this one works, i will use

Comment (Anhui-tqhuang) [reply pkg/utils/digits.go:20]: > Also the name isn't too descriptive, probably a better name will be "PaddingZero" or sometlhing like that that actually say what this function it's doing

i assume it's a bit different

this operation could call `PaddingZero`
```go
fmt.Sprintf("%0*d", len(fmt.Sprintf("%d", number)), i)
```

this operation could only call `IterativeDigitsCount`
```go
return len(fmt.Sprintf("%d", number))
```

File: internal/cmd/manager/instance/initdb/cmd.go
Comment (Anhui-tqhuang) [reply internal/cmd/manager/instance/initdb/cmd.go:116]: does it has access to secrets or configmaps, if so, it would easy, we even don't need to keep them in a folder, but if we want to keep them in a folder, anyway we need to know the folder name, except hardcoding the path, i cannot think of other ideas.

moreover the value of this option could be used to control the the logic
```go
func (info InitInfo) executePostInitApplicationSQLRefs(sqlUser *sql.DB) error {
	if info.PostInitApplicationSQLRefsFolder == "" {
		return nil
	}
        # ....
}
```

File: pkg/utils/digits.go
Comment (sxd) [inline pkg/utils/digits.go:22]: Now doesn't make sense to have this function in one file and one file to test this function, since it's an inliner it can be used directly when you need the value

File: pkg/specs/volumes.go
Comment (sxd) [inline pkg/specs/volumes.go:88]: here, we should use the inline code to get the count and will be easy to know how you get that number

File: pkg/utils/digits.go
Comment (Anhui-tqhuang) [reply pkg/utils/digits.go:22]: good point

File: docs/src/bootstrap.md
Comment (leonardoce) [inline docs/src/bootstrap.md:303]: I think we should say that we execute every SQL statement in a single exec on the server, and they will be executed according to the PostgreSQL semantics (described here https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-MULTI-STATEMENT).

This means, i.e., that the user can add comments to the script but cannot use any `psql` internal command.

**Review by leonardoce — CHANGES_REQUESTED** [review]

**Comment by leonardoce** [comment]
@Anhui-tqhuang this is awesome! Thank you!

File: pkg/specs/volumes.go
Comment (Anhui-tqhuang) [reply pkg/specs/volumes.go:109]: good point

File: .gitignore
Comment (Anhui-tqhuang) [reply .gitignore:33]: i have to vendor locally, otherwise GOLang server wouldn't work well

File: docs/src/bootstrap.md
Comment (Anhui-tqhuang) [reply docs/src/bootstrap.md:293]: understand, good point

**Comment by Anhui-tqhuang** [comment]
@leonardoce, thanks for the review i addressed comments, might be ready for another pass

File: docs/src/bootstrap.md
Comment (leonardoce) [reply docs/src/bootstrap.md]: 👍 could you please update the example inside `docs/src/samples/cluster-example-initdb-sql-refs.yaml` too? Thank you! 🙏

File: pkg/specs/jobs.go
Comment (leonardoce) [inline pkg/specs/jobs.go:35]: I think this can be private, as we use it just inside the `spec` package. I'm sorry to be so picky...

**Review by leonardoce — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (leonardoce) [reply .gitignore:33]: You could have added your rule inside `.git/info/exclude` instead of changing the versioned `.gitignore`.

File: pkg/specs/jobs.go
Comment (sxd) [reply pkg/specs/jobs.go:35]: should this go in `/etc/` ? probably we can put it in a directory inside `/controller`, I haven't check everything from scratch again so I may be forgetting sometlhing

Comment (leonardoce) [reply pkg/specs/jobs.go:35]: We already mount secrets in `/etc/superuser-secret` and `/etc/app-secret` so I don't see any downside on using `/etc` name here. Maybe we could devise a better name for that directory... I think `post-init-application-sql` it's better than `post-init-application-sql-refs`. There are no refs here but just contents.

Comment (sxd) [reply pkg/specs/jobs.go:35]: Yes, but those are configuration or secrets require for configuration, that's why `/etc` it's ok, but as you said this is content, that's why I think `/etc/` it's no the proper place for content file, by convention `/etc` has always contain configuration and files needed by configuration, on the other hand `/controller` contains files with all the needed by the instance manager, even file with content

Comment (Anhui-tqhuang) [reply pkg/specs/jobs.go]: fixed with the following action:
- make the const be private
- replace `post-init-application-sql-refs` with `post-init-application-sql  `

**Review by leonardoce — APPROVED** [review]

File: docs/src/bootstrap.md
Comment (armru) [inline docs/src/bootstrap.md:304]: Is this still true?
now we do sort.Strings(files)

`	
if err := fileutils.EnsureDirectoryExist(info.PostInitApplicationSQLRefsFolder); err != nil {
		return fmt.Errorf("could not find directory: %s, err: %w", info.PostInitApplicationSQLRefsFolder, err)
	}

	files, err := fileutils.GetDirectoryContent(info.PostInitApplicationSQLRefsFolder)
	if err != nil {
		return fmt.Errorf("could not get directory content from: %s, err: %w",
			info.PostInitApplicationSQLRefsFolder, err)
	}

	sort.Strings(files)

	for _, file := range files {
		sql, ioErr := fileutils.ReadFile(path.Join(info.PostInitApplicationSQLRefsFolder, file))
		if ioErr != nil {
			return fmt.Errorf("could not read file: %s, err; %w", file, err)
		}

		if err = info.executeQueries(sqlUser, []string{string(sql)}); err != nil {
			return fmt.Errorf("could not execute queries: %w", err)
		}
	}
`

**Review by armru — APPROVED** [review]

**Comment by leonardoce** [comment]
We run the E2e test manually on EDB's infrastructure as we'd like this work to reach main soon.
Everything was green.

**Pull request merged**
